source/concat.js

```diff
@@ -34,24 +32,12 @@ import toString from './toString';
  *      R.concat([], []); //=> []
  */
 var concat = _curry2(function concat(a, b) {
-  if (_isArray(a)) {
-    if (_isArray(b)) {
-      return a.concat(b);
-    }
-    throw new TypeError(toString(b) + ' is not an array');
-  }
-  if (_isString(a)) {
-    if (_isString(b)) {
-      return a + b;
-    }
-    throw new TypeError(toString(b) + ' is not a string');
-  }
   if (a != null && _isFunction(a['fantasy-land/concat'])) {
     return a['fantasy-land/concat'](b);
   }
   if (a != null && _isFunction(a.concat)) {
     return a.concat(b);
   }
-  throw new TypeError(toString(a) + ' does not have a method named "concat" or "fantasy-land/concat"');
+  return semigroup_concat(a, b);
```

Comment:
Two things: is `semigroup_concat` used elsewhere?  If not, shouldn't we internalize it here?

Second, and this will probably only work well if the answer to that it yes, we still have some expectations that while all sorts of things can be `concat`ed, don't we want to order our tests from most likely to least, in which case, I would prefer to check Arrays, Strings, and Objects before the fantasy-land check.